src/main/java/org/opensearch/knn/index/query/KNNWeight.java

```diff
@@ -97,6 +104,79 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
         if (filterWeight != null && filterIdsArray.length == 0) {
             return KNNScorer.emptyScorer(this);
         }
+        final Map<Integer, Float> docIdsToScoreMap = new HashMap<>();
+
+        /*
+         * The idea for this optimization is to get K results, we need to atleast look at K vectors in the HNSW graph
+         * . Hence, if filtered results are less than K and filter query is present we should shift to exact search.
+         * This improves the recall.
+         */
+        if (filterWeight != null && filterIdsArray.length <= knnQuery.getK()) {
```

Comment:
I remember we were talking about this. That is a great idea. This is an optimization for which we need to do experiments to find some candidate values like 1k or 500 when we should do exact search. I will take this as part of next optimization.

I want to call that as a ratio of filtered ids and K for which exact search yields faster results.

Reply:
just for my understanding, with current code will be doing search correctly, just less effectively, not dropping valid ids from result?

Reply:
my understanding is that for cases when after filtering we do have small subset of data, running HNSW on a full dataset is less effective (slower) than exact search on filtered subset. 
I want to understand if with this new code and in case that Vamshi has mentioned we may run search in a not optimal way but still getting correct nearest neighbors.

Reply:
@martin-gaievski 
Running HNSW on full data set with filtered Ids to be used for selection is both non-optimal in terms of latency and recall. Hence we want to switch to exact search in cases when filterIds size < k.

With new code we are doing exact search only if filterIds size < k. This is coming from the fact that to get K results from HNSW graph you have to look at atleast k nodes, now if the filtered Id set is already shrinked to < k nodes exact search is appropriate here. Hence it is optimal.

Now coming back to vamshi suggestion, its is a good suggestion but we cannot just rely on a magic number like 1k. My idea there is we should treat this as some ratio between k and filteredIds. But to do that we need to run experiments, which we can take incremental.

Reply:
Sounds good!  Lets create a Github issue to track this optimization if we plan to do incrementally.